rasa/nlu/training_data/training_data.py

```diff
@@ -427,16 +430,40 @@ def split_nlu_examples(
         self, train_frac: float, random_seed: Optional[int] = None
     ) -> Tuple[list, list]:
         train, test = [], []
-        for intent, count in self.examples_per_intent.items():
-            ex = [e for e in self.intent_examples if e.data[INTENT] == intent]
+        training_examples = set(self.training_examples)
+
+        def _split(_examples: List[Message], _count: int) -> None:
             if random_seed is not None:
-                random.Random(random_seed).shuffle(ex)
+                random.Random(random_seed).shuffle(_examples)
             else:
-                random.shuffle(ex)
+                random.shuffle(_examples)
+
+            n_train = int(_count * train_frac)
+            train.extend(_examples[:n_train])
+            test.extend(_examples[n_train:])
+
+        # to make sure we have at least one example per response and intent in the
+        # training/test data, we first go over the response examples and then go over
+        # intent examples
+
+        for response, count in self.examples_per_response.items():
+            examples = [
```

Comment:
optional: we can extract that to `_get_responses_by_name()` or similar.

Reply:
as we are not getting the actual responses, but the number of examples per response, I would either keep it or maybe rename to `number_of_examples_per_response`